pkg/pillar/canbus/canbus.go

```diff
@@ -0,0 +1,853 @@
+		case unix.IFLA_CAN_BITTIMING_CONST:
+			link.BitTimingConst.Tseg1_min = native.Uint32(datum.Value[16:])
+			link.BitTimingConst.Tseg1_max = native.Uint32(datum.Value[20:])
+			link.BitTimingConst.Tseg2_min = native.Uint32(datum.Value[24:])
+			link.BitTimingConst.Tseg2_max = native.Uint32(datum.Value[28:])
+			link.BitTimingConst.Sjw_max = native.Uint32(datum.Value[32:])
+			link.BitTimingConst.Brp_min = native.Uint32(datum.Value[36:])
+			link.BitTimingConst.Brp_max = native.Uint32(datum.Value[40:])
+			link.BitTimingConst.Brp_inc = native.Uint32(datum.Value[44:])
+		case unix.IFLA_CAN_DATA_BITTIMING:
+			link.DataBitTiming.Bitrate = native.Uint32(datum.Value)
+			link.DataBitTiming.Sample_point = native.Uint32(datum.Value[4:])
+			link.DataBitTiming.Tq = native.Uint32(datum.Value[8:])
+			link.DataBitTiming.Prop_seg = native.Uint32(datum.Value[12:])
+			link.DataBitTiming.Phase_seg1 = native.Uint32(datum.Value[16:])
+			link.DataBitTiming.Phase_seg2 = native.Uint32(datum.Value[20:])
+			link.DataBitTiming.Sjw = native.Uint32(datum.Value[24:])
+			link.DataBitTiming.Brp = native.Uint32(datum.Value[28:])
+		case unix.IFLA_CAN_DATA_BITTIMING_CONST:
+			link.DataBitTimingConst.Tseg1_min = native.Uint32(datum.Value[16:])
+			link.DataBitTimingConst.Tseg1_max = native.Uint32(datum.Value[20:])
+			link.DataBitTimingConst.Tseg2_min = native.Uint32(datum.Value[24:])
+			link.DataBitTimingConst.Tseg2_max = native.Uint32(datum.Value[28:])
+			link.DataBitTimingConst.Sjw_max = native.Uint32(datum.Value[32:])
+			link.DataBitTimingConst.Brp_min = native.Uint32(datum.Value[36:])
+			link.DataBitTimingConst.Brp_max = native.Uint32(datum.Value[40:])
+			link.DataBitTimingConst.Brp_inc = native.Uint32(datum.Value[44:])
+		case unix.IFLA_CAN_BERR_COUNTER:
+			link.ErrorCounters.Txerr = native.Uint16(datum.Value)
+			link.ErrorCounters.Rxerr = native.Uint16(datum.Value[2:])
+		case unix.IFLA_CAN_CLOCK:
+			link.Clock.Freq = native.Uint32(datum.Value)
+		case unix.IFLA_CAN_CTRLMODE:
+			link.CtrlMode.Mask = native.Uint32(datum.Value)
+			link.CtrlMode.Flags = native.Uint32(datum.Value[4:])
+		case unix.IFLA_CAN_STATE:
+			link.State = native.Uint32(datum.Value)
+		case unix.IFLA_CAN_RESTART_MS:
+			link.RestartMs = native.Uint32(datum.Value)
+		case unix.IFLA_CAN_RESTART:
+			link.Restart = native.Uint32(datum.Value)
+		case unix.IFLA_CAN_TERMINATION:
+			link.Termination = native.Uint16(datum.Value)
+		case IFLA_CAN_TDC | nl.NLA_F_NESTED:
+			nmsg, err := nl.ParseRouteAttr(datum.Value)
+			if err != nil {
+				continue
+			}
+			for _, nested := range nmsg {
+				switch nested.Attr.Type {
+				case IFLA_CAN_TDC_TDCV:
+					link.TDCV = native.Uint32(nested.Value)
+				case IFLA_CAN_TDC_TDCO:
+					link.TDCO = native.Uint32(nested.Value)
+				case IFLA_CAN_TDC_TDCF:
+					link.TDCF = native.Uint32(nested.Value)
+				}
+			}
+		}
+	}
+}
+
+// Set a bit flag
+func setBitFlag(flags uint32, mask uint32) uint32 {
```

Comment:
I usually tried to move such simple functions to the `pkg/pillar/utils` dir.

Reply:
I agree, I won't migrate in this case because I want to let this library 100% independent of pillar in order to merge it to netlink package and push it upstream in the future....